src/database/postgres_test.go

```diff
@@ -0,0 +1,253 @@
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+
+	err = cmd.Run()
+	if err != nil {
+		return fmt.Errorf("Failed to setup PostgreSQL server: %v", err)
+	}
+
+	return nil
+}
+
+func createTestingDatabase() (*pg.DB, error) {
+	// Connect to the default "postgres" database
+	db := pg.Connect(&pg.Options{
+		User:     "postgres",
+		Password: "postgres",
+		Addr:     "localhost:5432",
+		Database: "postgres",
+	})
+
+	// Drop the database if it exists, then create it
+	_, err := db.Exec("DROP DATABASE IF EXISTS testdb")
+	if err != nil {
+		db.Close()
+		return nil, fmt.Errorf("failed to drop existing test database: %v", err)
+	}
+
+	_, err = db.Exec("CREATE DATABASE testdb")
+	if err != nil {
+		db.Close()
+		return nil, fmt.Errorf("failed to create test database: %v", err)
+	}
+
+	// Close the connection to the "postgres" database
+	db.Close()
+
+	// Connect to the newly created database
+	testDB := pg.Connect(&pg.Options{
+		User:     "postgres",
+		Password: "postgres",
+		Addr:     "localhost:5432",
+		Database: "testdb",
+	})
+
+	return testDB, nil
+}
+
+func removeTestingDatabase() error {
+	// Connect to the default "postgres" database to drop the test database
+	defaultDB := pg.Connect(&pg.Options{
+		User:     "postgres",
+		Password: "postgres",
+		Addr:     "localhost:5432",
+		Database: "postgres",
+	})
+	defer defaultDB.Close()
+
+	// Terminate all connections to the test database
+	_, err := defaultDB.Exec("SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE datname = 'testdb'")
+	if err != nil {
+		return fmt.Errorf("Failed to terminate connections to test database: %v", err)
+	}
+
+	// Drop the test database
+	_, err = defaultDB.Exec("DROP DATABASE IF EXISTS testdb")
+	if err != nil {
+		return fmt.Errorf("Failed to drop test database: %v", err)
+	}
+
+	return nil
+}
+
+func TestPostgres(t *testing.T) {
+	const maxRetries = 3
+	var err error
+
+	if err := setupPostgres(); err != nil {
+		t.Fatalf("Failed to setup Postgres server: %v", err)
+	}
+
+	// Set a mock DATABASE_URL for testing
+	os.Setenv("DATABASE_URL", "postgres://postgres:postgres@localhost:5432/testdb")
+	defer os.Unsetenv("DATABASE_URL")
+
+	var db *pg.DB
+	for i := 0; i < maxRetries; i++ {
+		db, err = createTestingDatabase()
+		if err == nil || !strings.Contains(err.Error(), "connection reset by peer") {
+			break
+		}
+		t.Logf("Failed to create test database: Connection Reset by peer. Retrying...")
+		time.Sleep(time.Second * 2)
+	}
+	if err != nil {
+		t.Fatalf("Failed to create test database after %d attempts: %v", maxRetries, err)
+	}
+	defer db.Close()
+
+	// Test connection
+	ctx := context.Background()
+	conn, err := establishConnection(ctx)
+
+	assert.NoError(t, err)
+	defer conn.Close(ctx)
+
+	// Test database table initialisation
+	err = initializeDatabaseSchema(ctx, conn)
+	if err != nil {
+		t.Fatalf("Failed to initialise the database schema: %v", err)
+	}
+
+	// Test row insertion
+	err = testInsertRow(ctx, conn)
```

Comment:
These should be their own separate unit tests, not functions within this test.

Reply:
Done